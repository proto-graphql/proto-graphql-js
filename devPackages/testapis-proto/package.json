{
  "name": "@proto-graphql/testapis-proto",
  "version": "0.1.1",
  "description": "Generated codes from protobuf IDL",
  "type": "module",
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.js"
      },
      "require": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "sideEffects": false,
  "private": true,
  "scripts": {
    "build": "tsup",
    "prebuild": "./scripts/build-proto.ts"
  },
  "dependencies": {
    "@bufbuild/protobuf": "catalog:"
  },
  "devDependencies": {
    "@proto-graphql/tsconfig": "workspace:*",
    "glob": "^10.4.5"
  }
}
