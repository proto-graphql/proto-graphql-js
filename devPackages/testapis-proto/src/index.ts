import { create, fromBinary } from "@bufbuild/protobuf";
import {
  type CodeGeneratorRequest,
  CodeGeneratorRequestSchema,
  type FileDescriptorSet,
  FileDescriptorSetSchema,
} from "@bufbuild/protobuf/wkt";
import { fileDescriptorSetBins } from "./__generated__/fileDescriptorSetBins";

function objectKeys<K extends string>(obj: Record<K, unknown>): K[] {
  return Object.keys(obj) as K[];
}

export const testapisPackages = objectKeys(fileDescriptorSetBins);
export type TestapisPackage = keyof typeof fileDescriptorSetBins;

export function getTestapisFileDescriptorSet(
  pkg: TestapisPackage,
): FileDescriptorSet {
  return fromBinary(
    FileDescriptorSetSchema,
    Buffer.from(fileDescriptorSetBins[pkg], "base64"),
  );
}

export function buildCodeGeneratorRequest(
  pkg: TestapisPackage,
  { param }: { param?: string } = {},
): CodeGeneratorRequest {
  const descSet = getTestapisFileDescriptorSet(pkg);
  const req = create(CodeGeneratorRequestSchema, { parameter: param });

  for (const fd of descSet.file) {
    req.protoFile.push(fd);

    const filename = fd.name;
    const pat = new RegExp(`^${pkg.replace(/\./g, "/")}/[^/]+\\.proto$`);
    if (filename && pat.test(filename)) {
      req.fileToGenerate.push(filename);
    }
  }

  return req;
}
