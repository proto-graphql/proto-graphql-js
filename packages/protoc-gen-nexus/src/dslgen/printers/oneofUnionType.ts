import type { Registry } from "@bufbuild/protobuf";
import {
  type OneofUnionType,
  type SquashedOneofUnionType,
  compact,
  protobufGraphQLExtensions,
} from "@proto-graphql/codegen-core";
import { type Code, code, joinCode, literalOf } from "ts-poet";

import {
  type NexusPrinterOptions,
  fieldType,
  impNexus,
  nexusTypeDef,
} from "./util.js";

/**
 * @example
 * ```ts
 * export cosnt Oneof = unionType({
 *   name: "Oneof",
 *   // ...
 * })
 * ```
 */
export function createOneofUnionTypeCode(
  type: OneofUnionType | SquashedOneofUnionType,
  registry: Registry,
  opts: NexusPrinterOptions,
): Code {
  const typeOpts = {
    name: type.typeName,
    description: type.description,
    definition: code`(t) => {
      t.members(${joinCode(
        type.fields.map((f) => fieldType(f, opts)),
        { on: "," },
      )});
    }`,
    extensions: protobufGraphQLExtensions(type, registry),
  };
  return code`export const ${nexusTypeDef(type)} = ${impNexus(
    "unionType",
  )}(${literalOf(compact(typeOpts))});`;
}
