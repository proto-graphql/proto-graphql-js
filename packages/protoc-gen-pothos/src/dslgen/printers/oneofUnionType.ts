import type { Registry } from "@bufbuild/protobuf";
import {
  type OneofUnionType,
  type SquashedOneofUnionType,
  compact,
  protobufGraphQLExtensions,
} from "@proto-graphql/codegen-core";
import { type Code, code, literalOf } from "ts-poet";

import {
  type PothosPrinterOptions,
  fieldTypeRef,
  pothosBuilder,
  pothosRef,
} from "./util";

/**
 * @example
 * ```ts
 * export cosnt Oneof = builder.unionType("Oneof", {
 *   types: [...],
 *   // ...
 * })
 * ```
 */
export function createOneofUnionTypeCode(
  type: OneofUnionType | SquashedOneofUnionType,
  registry: Registry,
  opts: PothosPrinterOptions,
): Code {
  const typeOpts = {
    types: type.fields.map((f) => fieldTypeRef(f, opts)),
    description: type.description,
    extensions: protobufGraphQLExtensions(type, registry),
  };
  return code`
    export const ${pothosRef(type)} =
      ${pothosBuilder(type, opts)}.unionType(${literalOf(
        type.typeName,
      )}, ${literalOf(compact(typeOpts))});
  `;
}
