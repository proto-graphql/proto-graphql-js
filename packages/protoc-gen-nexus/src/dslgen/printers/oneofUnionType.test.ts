import { getTestapisFileDescriptorSet, TestapisPackage } from "@proto-graphql/testapis-proto";
import { 
  OneofUnionType,
  SquashedOneofUnionType,
  TypeOptions,
  defaultScalarMapping
} from "@proto-graphql/codegen-core";
import { createFileRegistry } from "@bufbuild/protobuf";
import { describe, expect, test } from "vitest";
import { createOneofUnionTypeCode } from "./oneofUnionType.js";
import type { NexusPrinterOptions } from "./util.js";

function generateOneofUnionTypeCode(packageName: TestapisPackage, typeNameInProto: string, oneofFieldName: string, options: NexusPrinterOptions): string {
  const typeOptions: TypeOptions = {
    partialInputs: false,
    scalarMapping: defaultScalarMapping,
    ignoreNonMessageOneofFields: false,
  };

  const descSet = getTestapisFileDescriptorSet(packageName);
  const registry = createFileRegistry(descSet)
  const descMsg = registry.getMessage(`${packageName}.${typeNameInProto}`);
  if (descMsg === undefined) {
    throw new Error(`Message ${typeNameInProto} not found in package ${packageName}`);
  }

  const descOneof = descMsg.oneofs.find(d => d.name === oneofFieldName)
  if (descOneof === undefined) {
    throw new Error(`Oneof field ${oneofFieldName} not found in message ${typeNameInProto} in package ${packageName}`);
  }

  const oneofType = new OneofUnionType(descOneof, typeOptions);

  const code = createOneofUnionTypeCode(oneofType, registry, options).toString();

  return code.toString();
}

function generateSquashedOneofUnionTypeCode(packageName: TestapisPackage, typeNameInProto: string, options: NexusPrinterOptions): string {
  const typeOptions: TypeOptions = {
    partialInputs: false,
    scalarMapping: defaultScalarMapping,
    ignoreNonMessageOneofFields: false,
  };

  const descSet = getTestapisFileDescriptorSet(packageName);
  const registry = createFileRegistry(descSet)
  const descMsg = registry.getMessage(`${packageName}.${typeNameInProto}`);
  if (descMsg === undefined) {
    throw new Error(`Message ${typeNameInProto} not found in package ${packageName}`);
  }

  const oneofType = new SquashedOneofUnionType(descMsg, typeOptions);

  const code = createOneofUnionTypeCode(oneofType, registry, options).toString();

  return code.toString();
}

describe("createOneofUnionTypeCode", () => {
  describe("google-protobuf", () => {
    const options: NexusPrinterOptions = {
      dsl: "nexus",
      protobuf: "google-protobuf" as const,
      importPrefix: "@testapis/google-protobuf",
      emitImportedFiles: false,
      fileLayout: "proto_file",
      filenameSuffix: ".nexus",
    };

    test("generates code for a simple oneof union", () => {
      const code = generateOneofUnionTypeCode("testapis.oneof", "OneofParent", "required_oneof_members", options);
      expect(code).toMatchSnapshot();
    });

    test("generates code for optional oneof union", () => {
      const code = generateOneofUnionTypeCode("testapis.oneof", "OneofParent", "optional_oneof_members", options);
      expect(code).toMatchSnapshot();
    });

    test("generates code for import squashed union", () => {
      const code = generateSquashedOneofUnionTypeCode("testapis.edgecases.import_squashed_union.pkg1", "SquashedOneof", options);
      expect(code).toMatchSnapshot();
    });
  });

  describe("protobufjs", () => {
    const options: NexusPrinterOptions = {
      dsl: "nexus",
      protobuf: "protobufjs" as const,
      importPrefix: "@testapis/protobufjs",
      emitImportedFiles: false,
      fileLayout: "proto_file",
      filenameSuffix: ".nexus",
    };

    test("generates code for a simple oneof union", () => {
      const code = generateOneofUnionTypeCode("testapis.oneof", "OneofParent", "required_oneof_members", options);
      expect(code).toMatchSnapshot();
    });

    test("generates code for optional oneof union", () => {
      const code = generateOneofUnionTypeCode("testapis.oneof", "OneofParent", "optional_oneof_members", options);
      expect(code).toMatchSnapshot();
    });

    test("generates code for import squashed union", () => {
      const code = generateSquashedOneofUnionTypeCode("testapis.edgecases.import_squashed_union.pkg1", "SquashedOneof", options);
      expect(code).toMatchSnapshot();
    });
  });
});