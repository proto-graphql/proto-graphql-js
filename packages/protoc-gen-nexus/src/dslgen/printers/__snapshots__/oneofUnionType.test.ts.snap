// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createOneofUnionTypeCode > google-protobuf > 'generates code for a simple oneof uni…' 1`] = `
"import { unionType } from "nexus";

export const OneofParentRequiredOneofMembers = unionType({
  "name": "OneofParentRequiredOneofMembers",
  "description": "Required. disallow not_set.",
  "definition": (t) => {
    t.members(OneofMemberMessage1, OneofMemberMessage2);
  },
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.required_oneof_members",
      "name": "required_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "required_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "required_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > google-protobuf > 'generates code for import squashed un…' 1`] = `
"import { unionType } from "nexus";

export const SquashedOneof = unionType({
  "name": "SquashedOneof",
  "definition": (t) => {
    t.members(OneofMessage1);
  },
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.edgecases.import_squashed_union.pkg1.SquashedOneof",
      "name": "SquashedOneof",
      "package": "testapis.edgecases.import_squashed_union.pkg1",
      "fields": [{
        "name": "msg1",
        "type": "testapis.edgecases.import_squashed_union.pkg1.OneofMessage1",
        "options": { "[graphql.object_type]": { "squashUnion": true } },
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > google-protobuf > 'generates code for optional oneof uni…' 1`] = `
"import { unionType } from "nexus";

export const OneofParentOptionalOneofMembers = unionType({
  "name": "OneofParentOptionalOneofMembers",
  "definition": (t) => {
    t.members(OneofMemberMessage1, OneofMemberMessage2);
  },
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.optional_oneof_members",
      "name": "optional_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "optoinal_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "optoinal_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > protobufjs > 'generates code for a simple oneof uni…' 1`] = `
"import { unionType } from "nexus";

export const OneofParentRequiredOneofMembers = unionType({
  "name": "OneofParentRequiredOneofMembers",
  "description": "Required. disallow not_set.",
  "definition": (t) => {
    t.members(OneofMemberMessage1, OneofMemberMessage2);
  },
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.required_oneof_members",
      "name": "required_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "required_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "required_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > protobufjs > 'generates code for import squashed un…' 1`] = `
"import { unionType } from "nexus";

export const SquashedOneof = unionType({
  "name": "SquashedOneof",
  "definition": (t) => {
    t.members(OneofMessage1);
  },
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.edgecases.import_squashed_union.pkg1.SquashedOneof",
      "name": "SquashedOneof",
      "package": "testapis.edgecases.import_squashed_union.pkg1",
      "fields": [{
        "name": "msg1",
        "type": "testapis.edgecases.import_squashed_union.pkg1.OneofMessage1",
        "options": { "[graphql.object_type]": { "squashUnion": true } },
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > protobufjs > 'generates code for optional oneof uni…' 1`] = `
"import { unionType } from "nexus";

export const OneofParentOptionalOneofMembers = unionType({
  "name": "OneofParentOptionalOneofMembers",
  "definition": (t) => {
    t.members(OneofMemberMessage1, OneofMemberMessage2);
  },
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.optional_oneof_members",
      "name": "optional_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "optoinal_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "optoinal_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;
